go-controller/pkg/ovn/host_subnet_allocator.go

```diff
@@ -0,0 +1,177 @@
+		if (ipv4Mode && utilnet.IsIPv4CIDR(subnet) && !foundIPv4) || (ipv6Mode && utilnet.IsIPv6CIDR(subnet) && !foundIPv6) {
+			if err := sna.MarkSubnetsAllocated(nodeName, subnet); err == nil {
+				klog.Infof("Valid subnet %v allocated on node %s", subnet, nodeName)
+				existingSubnets[n] = subnet
+				n++
+				if utilnet.IsIPv4CIDR(subnet) {
+					foundIPv4 = true
+				} else if utilnet.IsIPv6CIDR(subnet) {
+					foundIPv6 = true
+				}
+				continue
+			}
+		}
+		// this subnet is no longer needed; release it
+		klog.Infof("Releasing unused or invalid subnet %v on node %s", subnet, nodeName)
+		if err := sna.base.ReleaseNetworks(nodeName, subnet); err != nil {
+			klog.Warningf("Failed to release subnet %v on node %s: %v", subnet, nodeName, err)
+		}
+	}
+	// recreate existingSubnets with the valid subnets
+	existingSubnets = existingSubnets[:n]
+
+	// Node has enough valid subnets already allocated
+	if len(existingSubnets) == expectedHostSubnets {
+		klog.Infof("Allowed existing subnets %v on node %s", existingSubnets, nodeName)
+		return existingSubnets, allocatedSubnets, nil
+	}
+
+	// allocate new subnets if needed
+	if ipv4Mode && !foundIPv4 {
+		allocatedHostSubnet, err := sna.base.AllocateIPv4Network(nodeName)
+		if err != nil {
+			return nil, nil, fmt.Errorf("error allocating network for node %s: %v", nodeName, err)
+		}
+		// the allocator returns nil if it can't provide a subnet
+		// we should filter them out or they will be appended to the slice
+		if allocatedHostSubnet != nil {
+			klog.V(5).Infof("Allocating subnet %v on node %s", allocatedHostSubnet, nodeName)
+			allocatedSubnets = append(allocatedSubnets, allocatedHostSubnet)
+			// Release the allocation on error
+			defer func() {
+				if err != nil {
+					klog.Warningf("Releasing subnet %v on node %s: %v", allocatedHostSubnet, nodeName, err)
+					if errR := sna.base.ReleaseNetworks(nodeName, allocatedHostSubnet); errR != nil {
+						klog.Warningf("Error releasing subnet %v on node %s", allocatedHostSubnet, nodeName)
+					}
+				}
+			}()
+		}
+	}
+	if ipv6Mode && !foundIPv6 {
+		allocatedHostSubnet, err := sna.base.AllocateIPv6Network(nodeName)
+		if err != nil {
+			return nil, nil, fmt.Errorf("error allocating network for node %s: %v", nodeName, err)
+		}
+		// the allocator returns nil if it can't provide a subnet
+		// we should filter them out or they will be appended to the slice
+		if allocatedHostSubnet != nil {
+			klog.V(5).Infof("Allocating subnet %v on node %s", allocatedHostSubnet, nodeName)
+			allocatedSubnets = append(allocatedSubnets, allocatedHostSubnet)
+			// Release the allocation on error
+			defer func() {
+				if err != nil {
+					klog.Warningf("Releasing subnet %v on node %s: %v", allocatedHostSubnet, nodeName, err)
+					if errR := sna.base.ReleaseNetworks(nodeName, allocatedHostSubnet); errR != nil {
+						klog.Warningf("Error releasing subnet %v on node %s", allocatedHostSubnet, nodeName)
+					}
+				}
+			}()
+		}
+	}
```

Comment:
suggestion to clean up:
```
	allocate := func(nodeName string, allocator func(string ) (*net.IPNet, error)) {
		allocatedHostSubnet, err := allocator(nodeName)
		if err != nil {
			err = fmt.Errorf("error allocating network for node %s: %v", nodeName, err)
		}
		// the allocator returns nil if it can't provide a subnet
		// we should filter them out or they will be appended to the slice
		if allocatedHostSubnet != nil {
			klog.V(5).Infof("Allocating subnet %v on node %s", allocatedHostSubnet, nodeName)
			allocatedSubnets = append(allocatedSubnets, allocatedHostSubnet)
		}
	}

	// allocate new subnets if needed
	if ipv4Mode && !foundIPv4 {
		allocator := sna.base.AllocateIPv4Network
		allocate(nodeName, allocator)
	}

	if ipv6Mode && !foundIPv6 {
		allocator := sna.base.AllocateIPv6Network
		allocate(nodeName, allocator)
	}
```

Reply:
It's even easier than that:

```
	if ipv4Mode && !foundIPv4 {
		if err := allocateOneSubnet(sna.base.AllocateIPv4Network(nodeName)); err != nil {
			return nil, nil, err
		}
	}
```